### PR TITLE
feat(sdk): add interrupt mode to filesystem permissions

### PR DESCRIPTION
**Pull request by nick-hollon-lc**
## Example

```python
from deepagents import create_deep_agent
from deepagents.middleware.filesystem import FilesystemPermission

agent = create_deep_agent(
    model="anthropic:claude-sonnet-4-5",
    permissions=[
        # Pause for human approval on any write under /secrets/**
        FilesystemPermission(
            operations=["write", "edit"],
            paths=["/secrets/**"],
            mode="interrupt",
        ),
    ],
)
```

When the agent calls a filesystem tool whose access matches an interrupt-mode rule, the run pauses for approval via `HumanInTheLoopMiddleware` instead of erroring (`mode="deny"`) or running silently (`mode="allow"`). HITL is auto-installed whenever any rule uses `mode="interrupt"`.

## Summary

- Adds `mode="interrupt"` to `FilesystemPermission` — matching tool calls pause for human approval via `HumanInTheLoopMiddleware` instead of erroring out or running silently.
- `create_deep_agent` auto-installs HITL when any interrupt-mode rule is present and merges generated `interrupt_on` entries with the caller's `interrupt_on` (caller wins per tool name). Wired at all three sites: main agent, general-purpose subagent, and declarative subagents.
- Uses the `when` predicate on `InterruptOnConfig` (langchain-ai/langchain#37579), so interrupts fire per matching path/subtree rather than per tool name. The gate is **scope-aware** per tool, so it fires precisely without leaking protected paths.
- The interrupt offers the approver the full decision set — `approve` / `edit` / `reject` / `respond` — matching langchain's default for user-supplied `interrupt_on` tools.

## Implementation

- `deepagents/middleware/filesystem.py`
  - `FilesystemPermission.mode: Literal["allow", "deny", "interrupt"]`.
  - `_check_fs_permission` returns the matched rule's mode (first-match precedence — list a `deny` before an `interrupt` if you want deny to win).
  - Result-filtering helpers (`_filter_paths_by_permission`, `_filter_file_infos_by_permission`, `_filter_grep_matches_by_permission`) drop only `deny` matches; interrupt-mode paths remain in `ls`/`glob`/`grep` output because the HITL gate fires **before** execution (the predicate), so an approved bulk listing still returns its results.
- `deepagents/middleware/_fs_interrupt.py` (new) — the HITL glue, kept out of `FilesystemMiddleware` since only graph assembly needs it. `_FS_TOOL_PATH_ARGS` tags each tool with a `ToolScope`, and `_build_interrupt_on_from_permissions(rules)` returns a `{tool_name: InterruptOnConfig}` map with a scope-aware `when` predicate:
  - **exact** (`read_file` / `write_file` / `edit_file`) — fires iff the call's literal path matches an interrupt-mode rule.
  - **bulk** (`ls` / `glob` / `grep`) — fires iff the call's search subtree could intersect an interrupt-mode rule. A pathless call (`grep(path=None)`) and current-directory aliases (`.`, `""`, `./`) fire unconditionally because they can't be localized; `glob`'s `pattern` is also gated, since an absolute pattern (or `..` traversal) redirects the search root independently of `path`.
- `deepagents/backends/utils.py` — generic path helpers `_glob_anchor` (longest wildcard-free prefix of a glob) and `_paths_overlap` (component-aware subtree intersection), alongside `validate_path` / `to_posix_path`.
- `deepagents/graph.py` — merges fs-derived configs with `interrupt_on` and conditionally adds `HumanInTheLoopMiddleware` for the main agent, GP subagent, and declarative subagents.

## Security hardening (from review)

Corridor flagged, and this PR closes, three HITL-bypass shapes for bulk tools — each with a regression test:

- **Pathless `grep`** (`grep(path=None)`) — the predicate now fires unconditionally when the call can't be localized.
- **Current-directory aliases** (`.` / `""` / `./`, which `validate_path` normalizes to `/.`) — collapsed to `/` so the root-overlaps-everything branch fires.
- **Absolute / `..` glob `pattern`** (`glob(pattern="/secrets/**", path="/workspace")`) — Python's `glob` ignores the working directory for absolute patterns, so the predicate now gates `pattern` in addition to `path`.

## Dependency

This needs the `when` predicate / `interrupt_mode` on `HumanInTheLoopMiddleware`, which shipped in **langchain 1.3.3**. No git override or branch pin — the dependency uses the released package:

```toml
"langchain>=1.3.3,<2.0.0",
```

(The earlier `[tool.uv.sources]` override that pinned `langchain` to the `sr/hitl-refactor` branch has been removed now that langchain-ai/langchain#37579 is merged and released.)

The smoke-test snapshots track langchain's released tool descriptions (e.g. the `write_todos` description).

## Test plan

- [x] `make test` — 1727 passed, 90 skipped, 4 xfailed
- [x] `make lint` — clean (ruff + ty)
- [x] Unit tests in `tests/unit_tests/test_permissions.py` cover:
  - The `"interrupt"` literal on the dataclass.
  - `_check_fs_permission` returns `"interrupt"` on match and `deny`-before-`interrupt` precedence.
  - `_filter_paths_by_permission` keeps interrupt paths (only deny is filtered).
  - `_build_interrupt_on_from_permissions` registers only the tools whose op could be triggered.
  - The scope-aware `when` predicate: **exact** behavior, **bulk** subtree overlap, `_glob_anchor` / `_paths_overlap` edge cases (root, equal, ancestor, descendant, prefix lookalike, disjoint), and all three bypass regressions above (pathless, current-dir aliases, absolute/`..` glob patterns).

### CODE REVIEW

File: libs/deepagents/deepagents/middleware/filesystem.py [hunk 1]
```diff
@@ -182,6 +196,72 @@ def _apply_permissions_to_glob_results(
...
+    The predicate returns ``True`` iff the request's path argument resolves to
+    a rule with ``mode == "interrupt"`` for the given operation. Normal
+    first-match precedence applies, so a preceding ``deny`` rule wins and the
+    interrupt does not fire (the tool will return a permission-denied error).
+    """
+
+    def when(req: ToolCallRequest) -> bool:
+        raw_path = req.tool_call.get("args", {}).get(path_arg_name)
+        if not isinstance(raw_path, str):
+            return False
```
[the remark this answers is not in the document]
Comment (nick-hollon-lc) [reply libs/deepagents/deepagents/middleware/filesystem.py]: Valid finding, thanks. Fixed in f6880d55.

The three result-filter helpers (`_filter_paths_by_permission`, `_filter_file_infos_by_permission`, `_filter_grep_matches_by_permission`) now drop both `deny`- and `interrupt`-mode matches. The pre-execution `when` predicate is still load-bearing for the per-path tools (`read_file`/`write_file`/`edit_file`, which require an explicit path), but for the bulk tools (`ls`/`glob`/`grep`) the filter is now the authoritative gate — closing the pathless-`grep` hole and the parent-directory variant (e.g. `ls /` showing children of an interrupt-protected dir).

Added a regression test in `test_grep_path_none_does_not_leak_interrupt_paths`, plus updated the existing `_filter_paths_by_permission` assertion to reflect the new (strict) semantics.

Trade-off worth flagging: this means a user who set `mode="interrupt"` on `/secrets/**` and explicitly approves `ls /secrets` would still see an empty listing, because the post-filter can't tell that the interrupt fired and was approved. The intended workflow is now: use `read_file`/`write_file`/`edit_file` for interrupt-protected paths (each path triggers its own HITL prompt). Open to revisiting if a way to plumb "this call was approved" through to the filter becomes available.

[the remark this answers is not in the document]
Comment (nick-hollon-lc) [reply libs/deepagents/deepagents/middleware/filesystem.py]: Follow-up in 51b7a02d — the previous fix (f6880d55) closed the bypass but had a real UX regression I shouldn't have shipped: approving an interrupt-protected `ls`/`glob`/`grep` returned an empty result because the post-filter unconditionally stripped interrupt entries.

This commit moves the gate to where it belongs — the pre-execution `when` predicate — and reverts the filter helpers to stripping `deny` only.

`_FS_TOOL_PATH_ARGS` now tags each tool with a scope:

- **exact** (`read_file`/`write_file`/`edit_file`) — fires iff the literal path matches an interrupt-mode rule. Unchanged.
- **bulk** (`ls`/`glob`/`grep`) — fires iff the call's search subtree could intersect an interrupt-mode rule's pattern:
  - `grep(path=None)` → fires unconditionally (we can't localize the access).
  - explicit path → fires when the call subtree overlaps any interrupt rule's glob anchor. So `ls /` and `ls /secrets/sub` both fire under a rule on `/secrets/**`; `/workspace` and the prefix lookalike `/secret` do not.

This closes the original finding (pathless `grep` no longer bypasses HITL — it now triggers it) and the parent-directory variant (`ls /` with a rule on `/secrets/**` triggers HITL too), while restoring the approve→see-results UX.

New helpers `_glob_anchor` (extracts wildcard-free prefix) and `_paths_overlap` (component-aware subtree intersection) plus 14 unit tests covering both predicate branches and the overlap edge cases (root, equal, ancestor, descendant, prefix lookalike, disjoint).

File: libs/deepagents/deepagents/middleware/filesystem.py [hunk 2]
```diff
@@ -182,6 +206,162 @@ def _apply_permissions_to_glob_results(
...
+        raw_path = req.tool_call.get("args", {}).get(path_arg_name)
+        if raw_path is None:
+            # Pathless bulk call — fire because we cannot localize the access.
+            return True
+        if not isinstance(raw_path, str):
+            return False
+        try:
+            normalized = validate_path(raw_path)
+        except ValueError:
+            return False
```
[the remark this answers is not in the document]
Comment (nick-hollon-lc) [reply libs/deepagents/deepagents/middleware/filesystem.py]: Valid finding — fixed in 62d91bbe.

`validate_path(".")`, `validate_path("")`, and `validate_path("./")` all return `/.` (confirmed locally), which doesn't string-prefix any anchor, so the bulk predicate returned False and the call ran without HITL. The fix collapses normalized `/.` to `/` inside the bulk predicate so the root-overlaps-everything branch fires. A regression test now asserts every alias for "current/whole tree" (`.`, `""`, `./`, `/.`, `/`) triggers the interrupt.

On your follow-up hardening list — most don't apply to this codebase (deepagents' virtual filesystem has no symlinks, paths are POSIX-only, type validation happens upstream via the tool schema). The one that *was* worth flagging is glob-anchor behavior for unanchored patterns like `/**/secrets`: the anchor collapses to `/`, which means every bulk call fires the interrupt — conservative, but noisy. We've decided that over-firing is the right trade-off here (alternative was either silent under-gating or a much heavier "run, scan results, then maybe HITL" middleware), and documented this on `FilesystemPermission.mode` and `_glob_anchor` so users know to anchor their patterns (`/secrets/**`, `/projects/*/secrets/**`) for precise gating.

Audit logging of approved scopes is a good idea but out of scope for this PR — would file as a follow-up.

File: libs/deepagents/deepagents/middleware/filesystem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nick-hollon-lc) [reply libs/deepagents/deepagents/middleware/filesystem.py]: Thanks for the hardening checklist — addressing inline:

- **Empty/odd path inputs**: fixed in 62d91bbe (current-dir aliases `"."`, `""`, `"./"`, `/.` now collapse to `/` and fire the interrupt). Bytes/other non-str types are filtered out upstream by the tool schema (path args are typed `str`) so they never reach the predicate.
- **Canonicalization**: `validate_path` handles `.`/`..`/trailing slashes; the `/.` quirk above is now handled in-predicate. No symlinks exist in deepagents' virtual filesystem, so realpath resolution isn't applicable.
- **Case/separators**: paths in deepagents are POSIX-only and case-sensitive by design; not a concern here.
- **Glob anchors with leading/mid wildcards**: covered by the conservative root-fallback in `_glob_anchor`. Documented this and the over-firing trade-off so users anchor their patterns for precise gating. See reply on the new finding.
- **Rule precedence (deny beats interrupt)**: already enforced — `_check_fs_permission` is first-match, and the deny check at each tool runs before any HITL prompt. Covered by `test_deny_rule_takes_precedence_when_listed_first`.
- **Symlink crossings**: no symlinks in the virtual FS.
- **Audit logging of approved scopes**: good idea, deferred to a follow-up — not blocking this PR.

File: libs/deepagents/deepagents/middleware/filesystem.py [hunk 3]
```diff
@@ -182,6 +213,173 @@ def _apply_permissions_to_glob_results(
...
+
+    For ``/secrets/**`` returns ``/secrets``; for ``/a/*/b`` returns ``/a``;
+    for a wildcard at the root (``/**/secrets``, ``/*/foo``) falls back to
+    ``/``, which causes the bulk predicate to fire for *any* bulk call —
+    conservative over-gating, since we cannot statically pin down where the
+    rule could resolve. Users wanting precise gating should anchor the
+    rule's leading components. Mirrors the backslash-normalization used by
+    :meth:`FilesystemPermission.__post_init__`.
+    """
+    norm = pattern.replace("\\", "/")
```
Comment (eyurtsev) [inline libs/deepagents/deepagents/middleware/filesystem.py]: is this correct? is there pathlib.Posix that we should be using so we don't mess stuff up?

File: libs/deepagents/tests/unit_tests/test_permissions.py [hunk 4]
```diff
@@ -164,6 +168,171 @@ def test_backslash_path_without_traversal_accepted(self):
...
+        "current/whole tree" as unlocalized.
+        """
+        rule = FilesystemPermission(operations=["read"], paths=["/secrets/**"], mode="interrupt")
+        when = _make_fs_when_predicate([rule], "read", "path", "bulk")
+        for alias in (".", "", "./", "/.", "/"):
+            assert when(_FakeReq({"path": alias})) is True, f"alias {alias!r} should fire"
+
+
+class TestGlobAnchorAndOverlap:
+    def test_glob_anchor_strips_double_star(self):
```
Comment (eyurtsev) [inline libs/deepagents/tests/unit_tests/test_permissions.py]:
```python
@pytest.parameterize(
)
```

**Review by sydney-runkle — APPROVED** [review]
awesome, let's have eugene review as well before merge

File: libs/deepagents/deepagents/middleware/_fs_interrupt.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    rules: list[FilesystemPermission],
+    operation: FilesystemOperation,
+    path_arg_name: str,
+) -> Callable[[ToolCallRequest], bool]:
+    def when(req: ToolCallRequest) -> bool:
+        raw_path = req.tool_call.get("args", {}).get(path_arg_name)
+        if not isinstance(raw_path, str):
+            return False
+        try:
+            normalized = validate_path(raw_path)
```
Comment (eyurtsev) [inline libs/deepagents/deepagents/middleware/_fs_interrupt.py:85]: (not related to this PR)

`validate_path` is confusing -> doing both validation and returns a value

`validate_` named functions in python usually only return None OR throw an exception

**Review by eyurtsev — APPROVED** [review]
